Paper: Python Array API Standard: Toward Array Interoperability in the Scientific Python Ecosystem

asmeurer opened the pull request:
Note that the code to run and generate the plots for the benchmarks in the paper is at https://github.com/data-apis/scipy-2023-presentation/tree/main/benchmarks.

If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

CC @kgryte @rgommers @leofang @honno @shoyer @amueller @szha @saulshanabrook @steff456 @lezcano @thomasjpfan @tylerjereddy @alextp @HyukjinKwon @teoliphant

*Editor*: <!--editor-->@cbcunc<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@jpivarski<!--end-reviewers-list-->

asmeurer committed:
Paper: Python Array API Standard: Toward Array Interoperability in the Scientific Python Ecosystem

deniederhut commented:
@scoobies assign @mepa as editor

scoobies commented:
Assigned! @mepa is now the editor

mepa commented:
@scoobies build status

scoobies commented:
{"asmeurer-array-api"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpbmx4iovf'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: aaron_meurer\n*** Warning: /tmp/tmpbmx4iovf/publisher/../output/aaron_meurer/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/asmeurer-array-api.pdf", "build_status"=>"success", "build_timestamp"=>"02/06 21:59"}, "status"=>"success"}}

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-4af1f417-00a is OK
- 10.48550/arXiv.1209.5145 is OK
- 10.1137/141000671 is OK
- 10.48550/arXiv.1512.01274 is OK
- 10.1145/42288.42291 is OK
- 10.1145/77626.79170 is OK
- 10.1063/1.4822400 is OK
- 10.1109/MCSE.2007.51 is OK
- 10.1007/s00355-011-0603-9 is OK
- 10.1080/10361146.2014.900530 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.5334/jors.148 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1109/IEEESTD.2019.8766229 is OK
- 10.3233/978-1-61499-649-1-87 is OK
- 10.1145/355841.355847 is OK
- 10.21105/joss.01891 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.1002/0471725293 is OK
- 10.1109/MCSE.2011.36 is OK
- 10.1145/3386331 is OK
- 10.1109/MCSE.2007.58 is OK
- 10.1109/MCSE.2007.53 is OK
- 10.1109/MCSE.2010.119 is OK
- 10.1201/9781315373461-6 is OK
- 10.25080/Majora-7b98e3ed-013 is OK
- 10.25080/Majora-92bf1922-011 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.7717/peerj.453 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1109/TAU.1967.1161901 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.5555/3026877.3026899 is INVALID
- 10.1.1.112.9899 is INVALID
- 10.5555/3454287.3455008 is INVALID
- 10.5555/1953048.2078195 is INVALID
```

asmeurer committed:
Add Travis as an author

cbcunc commented:
@scoobies help

scoobies commented:
Hello @cbcunc, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Add to this issue's reviewers list
@scoobies add @username to reviewers

# Remove from this issue's reviewers list
@scoobies remove @username from reviewers

# Assign a user as the editor of this submission
@scoobies assign @username as editor

# Remove the editor assigned to this submission
@scoobies remove editor

# Add a user to this issue's assignees list
@scoobies add assignee: @username

# Remove a user from this issue's assignees list
@scoobies remove assignee: @username

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from branch custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: needs-review. Remove labels: unready, needs-more-review, pending-comment, ready
@scoobies mark needs review

# Label issue with: needs-more-review. Remove labels: unready, needs-review, pending-comment, ready
@scoobies mark needs more review

# Label issue with: pending-comment. Remove labels: unready, needs-review, needs-more-review, ready
@scoobies mark pending comment

# Label issue with: ready. Remove labels: unready, needs-review, needs-more-review, pending-comment
@scoobies mark ready

# Label issue with: unready. Remove labels: needs-review, needs-more-review, pending-comment, ready
@scoobies mark not ready

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

cbcunc commented:
@scoobies remove editor

scoobies commented:
Editor removed!

cbcunc commented:
@scoobies assign @cbcunc as editor

scoobies commented:
Assigned! @cbcunc is now the editor

deniederhut commented:
@scoobies add @jpivarski to reviewers

scoobies commented:
@jpivarski added to the reviewers list!

jakirkham commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+:author: Travis E. Oliphant
+:email: travis@quansight.com
+:institution: Quansight
+
```
```suggestion

:author: John Kirkham
:email: jkirkham@nvidia.com
:equal-contributor:
:institution: NVIDIA Corporation

```

xref: https://github.com/data-apis/scipy-2023-presentation/pull/25

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+   interfaces (APIs) and behaviors for array and tensor objects and operations
+   as commonly found in libraries such as NumPy :cite:`Harris2020a`, PyTorch
+   :cite:`Paszke2019a`, TensorFlow :cite:`Abadi2016a`, Dask :cite:`Rocklin2015a`,
+   and CuPy :cite:`Okuta2017a`. The establishment and subsequent adoption of the
```
Here's a copy-edit thing: there are several lists of array library names in the paper. They are necessarily different lists, but they could at least be put in a common order to make it easier to compare them.

The section highlighted above, in the abstract, are just examples of libraries. There probably wasn't any criterion for what is included or excluded other than likelihood that a reader of the paper would be familiar with them.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L323-L324

has libraries that were consulted as prior art in the design phase.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L744-L745

has libraries in the compatibility layer.

The line

https://github.com/scipy-conference/scipy_proceedings/blob/98ebaee556e0c3c4c78d580e072047db0a3eee77/papers/aaron_meurer/paper.rst?plain=1#L679-L681

has libraries that are in the process of adopting the standard.

Fig. 2 and discussion has libraries used in the scikit-learn and SciPy tests. A plot is harder to change than a list order in text; the plot's order is PyTorch followed by CuPy.

jpivarski replied:
Maybe also the Index Terms (mostly a list of library names).

kgryte replied:
This has now been addressed. We've updated the content to use a consistent ordering for referenced libraries.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+until the founding of the Consortium.
+
+The genesis for the Consortium grew out of many conversations among maintainers
+during 2019-2020. During those conversations, it quickly became clear that any
```
```suggestion
during 2019‒2020. During those conversations, it quickly became clear that any
```

(n-dash)

kgryte replied:
This has been addressed. Thanks for catching!

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+deferred execution and operator fusion, functional APIs are generally preferred.
+This mirrors design patterns used in underlying implementations, such as those
+written in C/C++ and Fortran, and more closely matches written mathematical
+notation.
```
I found this use of the words "pure functions" and "functional API" to be confusing. As per [Wikipedia](https://en.wikipedia.org/wiki/Functional_programming):

> Functional programming is sometimes treated as synonymous with [purely functional programming](https://en.wikipedia.org/wiki/Purely_functional_programming), a subset of functional programming which treats all functions as [deterministic](https://en.wikipedia.org/wiki/Deterministic_system) mathematical [functions](https://en.wikipedia.org/wiki/Function_(mathematics)), or [pure functions](https://en.wikipedia.org/wiki/Pure_function).

I was expecting these words to mean that the API is designed to be side-effect free, but the text of this paragraph suggests that you mean free-standing functions, as opposed to methods (or binary operators?).

This point would be clearer if you replace the words "pure functions" and "functional API" with something else, maybe "unbound functions, rather than methods".

kgryte replied:
This has now been addressed. We replaced "functional" and "pure functions" with "standalone functions".

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+devices, such as graphics processing units (GPUs), tensor processing units
+(TPUs), and field-programmable gate arrays (FPGAs).
+
+**Compiler support.** Standardized APIs and behaviors should be amenable to
+just-in-time (JIT) and ahead-of-time (AOT) compilation and graph-based
+optimization techniques, such as those used by PyTorch :cite:`Paszke2019a`, JAX
+:cite:`Bradbury2018a`, and TensorFlow :cite:`Abadi2016a`. APIs and behaviors
+not amenable to compilation, such as APIs returning arrays having
+data-dependent output shapes, should be specified as optional. Moreover,
```
It's not, strictly speaking, compiler support that's at issue here—the issue is that shapes (just `ndim` or the whole `shape` tuple) are considered part of the data type that gets compiled in, right?

What you want to allow for is libraries that use compile-time literal shapes to more effectively vectorize and unroll loops (and maybe more advanced graph optimization techniques than I know about). For example, Numba knows `ndim` but not `shape` at compile-time and therefore is able to return data-dependent output shapes, but JAX knows the `shape` at compile-time and therefore is not able to (but compiles to faster code).

If so, then maybe it would be better to say here that you want to allow compilers to have any amount of array metadata fields at compile-time. Later, you define an array object's metadata fields as "data type," "shape," and "device," and that's the maximum amount of information you want to allow compilers to have.

rgommers replied:
The "compiler support" point is broader than shapes. Other examples include avoiding functions that are polymorphic (e.g. `numpy.unique` has a number of return values that depend on input keywords) or enforcing output dtype stability (no value-based casting allowed).

> Later, you define an array object's metadata fields as "data type," "shape," and "device," and that's the maximum amount of information you want to allow compilers to have.

Yes, that sounds right for arrays that get returned.

To address this comment, how about adding one or both examples above for compiler support explicitly, and then adding explicitly that dtype, shape and device of any returned arrays should be predictable from the input types/dtypes/shapes/devices only?

jpivarski replied:
That sounds like a good solution to me. The comment was about my wrong guess about what "compiler support" entails; illustrating with examples is a good way to do it.

kgryte replied:
This has now been addressed. We added the example of "return type polymorphism" and a sentence on return value predictability based on input arguments.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+
+To understand usage patterns of array libraries within the ecosystem, we first
+identified a representative sample of commonly used Python libraries
+("downstream libraries") which consume the aforementioned array libraries. The
+sample of downstream libraries included SciPy :cite:`Virtanen2020a`, pandas
+:cite:`McKinney2011a`, Matplotlib :cite:`Hunter2007a`, xarray
+:cite:`Hoyer2017a`, scikit-learn :cite:`Pedregosa2011a`, statsmodels
+:cite:`Seabold2010a`, and scikit-image :cite:`Vanderwalt2014a`, among others.
```
This is not a major point, but it seems like it would be possible to publish a complete list of the downstream libraries used to establish usage. You're being very formal with the voting systems that treat their usage fairly (Dowdall and Borda, referenced below), maybe there should also be a complete list of the constituency somewhere?

How large is it? A lot more than these 7 or just enough more that you don't want to use up page space by writing them all?

saulshanabrook replied:
[Here](https://github.com/data-apis/python-record-api/tree/master/data/api) is a list of all the libraries used to create usage data I believe, unless some were added/removed outside of the recording repo.

asmeurer replied:
As Saul noted, the full list of libraries is 

```
alphalens
hvplot
modin
pandas
skimage
xarray
dask
koalas
networkx
prophet
scipy
sklearn
geopandas
matplotlib
orange3
pyjanitor
seaborn
statsmodels
```

which is 18 libraries. We primarily only listed the most widely used libraries here for the sake of space. Should we remove "among others" here and replace it with a full list?

Also, hopefully it is clear that even though we used a formal voting system in this data analysis, this was only used to assign priorities for standardization. At the end of the day the final selection of which APIs to include was made by the Consortium.

kgryte replied:
As mentioned in other comments, we explicitly listed the most widely used libraries in the interest of space. We link (via the bibliography) to the Python record API repository where a more complete list can be found.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+support within the ecosystem may be considered candidates for standardization.
+
+
+Methods
+=======
```
Perhaps

```suggestion
Methodology
===========
```

because "methods" are part of an API, and that's not what you mean here.

kgryte replied:
This has now been addressed. Thanks for the suggestion!

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+To interact with array objects, one uses "indexing" to access sub-arrays and
+individual elements, "operators" to perform logical and arithmetic operations
+(e.g., `+`, `-`, `*`, `/`, and `@`), and array-aware functions (e.g., for
+linear algebra, statistical reductions, and element-wise computation). Array
+indexing semantics extend built-in Python sequence `__getitem__()` indexing
+semantics to support element access across multiple dimensions (`Fig. 1b`_).
```
Are in-place mutation semantics included? That is, `__setitem__`?

At one point, I thought that in-place mutation was going to be an optional extension, since some array libraries (e.g. JAX again) benefit by excluding it.

Maybe you really did mean "immutable data"/"no side effects" in the "pure functions"/"functional API" part of the Design Principles section—in which case, I misunderstood it!

asmeurer replied:
In-place semantics are supported, but they are only supported in situations where mutations do not affect more than one array (i.e., the behavior is undefined in the presence of "views"). See https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html#copyview-mutability. There's actually a lengthy discussion about this happening at https://github.com/data-apis/array-api/issues/609 if you are interested.

asmeurer replied:
In-place semantics are supported, but they are only supported in situations where mutations affect more than one array (i.e., the behavior is undefined in the presence of "views"). See https://data-apis.org/array-api/latest/design_topics/copies_views_and_mutation.html#copyview-mutability. There's actually a lengthy discussion about this happening at https://github.com/data-apis/array-api/issues/609 if you are interested.

rgommers replied:
> they are only supported in situations where mutations affect more than one array

Correction: only when _only one array_ is affected. I.e., at that point there is no difference between in-place and out-of-place versions of the same operation. Hence there is no issue with libraries implemented under the hood as immutable data structures (JAX, TensorFlow).

kgryte replied:
This has now been addressed. We added a footnote concerning in-place operations.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+       return xp.sin(y) * x
+
+To interact with array objects, one uses "indexing" to access sub-arrays and
+individual elements, "operators" to perform logical and arithmetic operations
+(e.g., `+`, `-`, `*`, `/`, and `@`), and array-aware functions (e.g., for
+linear algebra, statistical reductions, and element-wise computation). Array
+indexing semantics extend built-in Python sequence `__getitem__()` indexing
+semantics to support element access across multiple dimensions (`Fig. 1b`_).
+Indexing an array using a boolean array (also known as "masking") is an
+optional standardized behavior. The result of a mask operation is
+data-dependent and thus difficult to implement in array libraries relying on
+static analysis for graph-based optimization.
```
This is referring to the "compiler support" part of the Design Principles section: the fact that the `shape` array metadata must be known at compile-time.

That makes sense for indexing using a boolean array, but what about indexing using an integer array, which is more powerful and it does not require shape information to be derived from data? That is, in

```python
output_array = base_array[integer_array]
```

the `output_array.shape == integer_array.shape`.

I'm not saying that the standard needs this; I'm saying that the paper has left me wondering about it, since I usually think of slicing by boolean array and slicing by integer array as going hand-in-hand. (They're presented side-by-side in the same [section of the NumPy documentation](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing).) A short sentence saying that integer array indexing is included or why it's not would be enough.

asmeurer replied:
Integer array indexing wasn't omitted because of compiler support, but rather because of lack of broad library support, and because of caution surrounding implementing the full NumPy indexing behavior, which includes some poor design decisions that we don't want to encode into the spec. It is possible that a future version of the standard will include some subset of integer array indexing. 

There have been discussions about adding equivalent functional APIs, and the 2022.12 version of the spec includes the `take` function, which implements the equivalent of integer array getitem indexing for 1D arrays. The setitem equivalent, `put`, is still being discussed. 

See https://github.com/data-apis/array-api/issues/177 for the full discussions here.

kgryte replied:
This has now been addressed. We added a note concerning integer array indexing and the possibility of its inclusion in a future revision of the array API standard.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+Fourier transform APIs. To accommodate specialized APIs, the Python array API
+standard includes standardized optional extensions.
+
+An extension is defined as a coherent set of standardized functionality which
+is commonly implemented across many, but not all, array libraries. Due to
```
The format of an extension is not limited to a set of functions (e.g. the Fourier analysis functions in your motivating example), so it _could_ be something like defining `strides` as an array metadata field and what happens to the strides in indexing and other functions, right?

But then, I suppose that this extension would need to have something to say about all other extensions: how do strides interact with Fourier functions? That would make it a special extension, and if there are too many special extensions, all commenting on each other, that could become infeasible.

Does the definition of an extension exclude that possibility? Like, if an extension can only be an additional suite of functions, then this sentence could clarify that.

asmeurer replied:
Right now an "extension" in the standard is defined specifically as a sub-namespace https://data-apis.org/array-api/latest/extensions/index.html. This is distinct from "optional behavior" such as boolean indexing. Will make this clearer.

kgryte replied:
This has now been addressed. We clarified that an extension is a sub-namespace of related functionality.

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+Allow array-consuming libraries to accept and operate on arrays
+from multiple different array libraries. 2) Establish a common set of
+standardized APIs and behaviors, enabling more sharing and code reuse. 3) For
+new array libraries, offer a concrete API that can be adopted as is. 4)
```
```suggestion
new array libraries, offer a concrete API that can be adopted as-is. 4)
```

kgryte replied:
This has been addressed. Thanks for catching!

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+.. [#] To ensure that observed performance is not an artifact of the generated dataset, we tested performance across multiple random datasets and did not observe a measurable difference across benchmark runs.
+
+`Fig. 2a`_ and `Fig. 2b`_ display results, showing average execution time
+relative to NumPy. When fitting an LDA model (`Fig. 2a`_), we observe 1.9x
+higher throughput for PyTorch CPU, 7.9x for CuPy, and 45.1x for PyTorch GPU.
+When predicting class labels (`Fig. 2b`_), we observe 2.5x higher throughput
+for PyTorch CPU, 24.6x for CuPy, and 44.9x for PyTorch GPU. In both benchmarks,
```
```suggestion
relative to NumPy. When fitting an LDA model (`Fig. 2a`_), we observe 1.9×
higher throughput for PyTorch CPU, 7.9× for CuPy, and 45.1× for PyTorch GPU.
When predicting class labels (`Fig. 2b`_), we observe 2.5× higher throughput
for PyTorch CPU, 24.6× for CuPy, and 44.9× for PyTorch GPU. In both benchmarks,
```

kgryte replied:
This has been addressed. Thanks!

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+
+`Fig. 2c`_ and `Fig. 2d`_ display results, showing average execution time
+relative to NumPy. When using library-specific optimizations (`Fig. 2c`_), we
+observe 1.4x higher throughput for PyTorch CPU, 51.6x for PyTorch GPU, and
+52.4x for CuPy. When omitting library-specific optimizations (`Fig. 2d`_), we
+observe a 12-25x **decreased** throughput across all non-NumPy backends.
```
```suggestion
observe 1.4× higher throughput for PyTorch CPU, 51.6× for PyTorch GPU, and
52.4× for CuPy. When omitting library-specific optimizations (`Fig. 2d`_), we
observe a 12‒25× **decreased** throughput across all non-NumPy backends.
```

kgryte replied:
This has now been addressed. Thanks!

jpivarski commented on papers/aaron_meurer/paper.rst:
```diff
@@ -0,0 +1,1028 @@
+pipelines. We expect these tools to be particularly valuable to array-consuming
+libraries in order to quickly assess API portability.
+
+**Coordination**: Providing a forum for coordination among array libraries
+(and their consumers) was the primary motivating factor behind Consortium
+formation and is the most important byproduct of Consortium efforts. By
```
It seems like this would be a good place to point readers to your communications channels. Is it all on https://github.com/data-apis/array-api, or are there Discourse/Discord sites for people to join?

kgryte replied:
This has now been addressed. We added a note referring readers to the public issue tracker.

jpivarski reviewed (commented):
This is a very well-organized and thoughtful paper. I've read it through once, and I know that I won't have any (or many) line-by-line copy-edit suggestions.

For large-scale comments, I'll have to be careful to comment on what is or is missing from the paper, rather than comment directly on the API decisions. (There are other channels for that.) However, if something about the API decisions strikes me as odd, it could make sense to address it in the paper, since other readers might have the same questions.

The thing I was most surprised about in the read-through was that "strides" are not considered part of the array object. I recognize that some libraries (e.g. JAX) are able to do great things by not allowing strides to be free. But then, you show in the SciPy benchmark (Fig. 2c, 2d) that not having this handle implies a severe performance penalty, which is likely to be a major take-away from this paper by most readers. In the discussion, you say that it illustrates that users will need to apply library-specific optimizations after all. (Implementing a backend or runtime switching system is non-goal number 2.) It seems to me like strides-awareness would be a good candidate for one of the optional extensions, if extensions are not strictly sets of functions. Perhaps that's a question or suggestion I should take up with the Consortium's normal channels, not a review of the paper, but it's a question the paper leads me to have.

Another thing that seems like an omission is any discussion of error handling. Libraries that allow asynchronous computation (such as on a GPU or distributed by Dask) either need to raise errors that depend on values at a different time or not raise errors at all. CuPy, at least, treats out-of-bounds advanced indexing as not an error; it wraps indexes instead. The boundary between not-error and is-error is a part of an API. Does the array API define it or is that considered a non-goal? If the latter, one sentence would clarify that.

jpivarski commented:
I think I already said this, but I wanted to put a comment at the end of all the nit-picky callouts that this is a wonderfully organized, very clear paper, and a great project!

asmeurer commented:
Hi @jpivarski. Thank you for your thoughtful review. We'll be taking a closer look soon, but I wanted to ask what our deadline is for addressing these review comments. Also it would help if you could confirm whether your review is completed or if you may still have additional comments.

jpivarski commented:
Oh, I'd say the review is complete, apart from any iterating on comments. We were given until June 23 to write an initial review, but I'll be on vacation, getting back to work on June 26, so I made sure to do it before I left. You don't need to respond on a short timescale! We (the authors and reviewers) have until July 14 to finish iterating.

kgryte committed:
Add John Kirkham as an author

Co-authored-by: jakirkham <jakirkham@gmail.com>

kgryte committed:
Update author order

This order matches the order found in the upstream repository.

rgommers commented:
> The boundary between not-error and is-error is a part of an API.

I don't completely agree with that. It's part of any _implementation_ but not necessarily an _interface_. Because there's an infinite amount of ways in which you can deviate from a specified interface, and it's impossible to exhaustively specify what will happen in such cases.

Also think about other standards like the C/C++ language ones - they specify what is valid and must/should be done, but rarely how to deal with invalid inputs.

>  Does the array API define it or is that considered a non-goal? If the latter, one sentence would clarify that.

The standard does specify expected inputs and also in quite a few cases that an exception must be raised. It tries to avoid specifying what exception type must be raised, or at what point in the execution - because as you pointed out it is often implementation-dependent, with constraints imposed by hardware or programming model, whether such specs can be adhered to.

The draft 2023 version of the standard has a design page on exceptions: https://data-apis.org/array-api/draft/design_topics/exceptions.html.

rgommers commented:
The points/questions around strides are quite interesting, thanks @jpivarski. It'll be useful to have an in-depth design discussion on that very soon, probably in a new issue on https://github.com/data-apis/array-api/issues. For the purposes of this paper, I suspect that it could be enough to add strides handling, and maybe 1-2 other key points (device support comes to mind), as concrete focus points in Future Work. It fits well with this part:

_Adoption: In order to ensure the continued success and relevance of the Python array API standard, we work closely with maintainers of array and array-consuming libraries to facilitate adoption by soliciting feedback, addressing pain points,"_

I think that's basically where we're at now - libraries like scikit-learn and SciPy are starting to implement support and finding pain points, and those will be the top prios to resolve via a compat package and/or further standardization efforts.

jpivarski commented:
I was thinking of language specifications. The C and C++ specifications define the boundaries of allowed inputs by saying "such-and-such a situation is undefined" and the Java specification names the type of exception that should be raised. The reason that I thought of it is because the difference between NumPy and CuPy on what they do about out-of-bounds advanced indexes is a pain point, though I understand why CuPy does it: the error could only be discovered asynchronously on a coprocessor, which would make it very difficult to raise errors.

Thanks for the link! It answers my question: the specification has a "Raises" section for each function (maybe including `__getitem__`) with "should" rather than "must" specifications. Maybe a short note on that could be worked into the paper. My intention here shouldn't be to review the specification itself—it's already received a lot of attention—but the paper, so my comment should be understood as, "There was no mention of exceptions in the paper. Since you have a decision on what to do about exceptions, maybe just make a note so that readers of the paper aren't left wondering."

Thanks!

---------

Yes, just a short comment on strides in the Future Work section would be best. That's not even a commitment that there will be an optional extension for strides, just that it's recognized as a pain point. As a reader, it looked like a problem was revealed without an intention to address it; mentioning it in Future Work makes it clear that it's not being ignored.

leofang commented:
Regarding strides: I have conflicting views and will continue to self-debate. On the one hand, when over half of the array libraries (Jax, TF, Dask, cuNumeric) do not offer strides for various reasons (ex: it makes no sense for distributed arrays), it's my opinion that even making it optional is not the right way to go, just like many other design considerations that we've made. Also, for those supporting the stride semantics, their strides are accessible via DLPack, it's just that currently it's hard to access them from within pure Python. On the other hand, as a low-level C++ library developer (in short, we write backends for the array libraries), I do see the pain point that it's hard for us to even imagine how our library would be adopted and consumed by the array libraries that do not have strides. It could be possible that our library simply cannot serve as the battery behind them due to different design philosophies, idk. Though, to be fair, the standard aims at the array libraries, not their backends, so if there's a standardization effort targeting array backends, it might be a better place for standardizing strides.

rgommers commented:
I opened https://github.com/data-apis/array-api/issues/641 to summarize and continue the conversation around strides.

cbcunc commented:
@asmeurer We are beyond the final review deadline. Please address the review comments and resolve them. Thank you.

asmeurer commented:
@cbcunc what is/was our final review deadline? I [asked about this above](https://github.com/scipy-conference/scipy_proceedings/pull/822#issuecomment-1592121021), and [the response I got](https://github.com/scipy-conference/scipy_proceedings/pull/822#issuecomment-1592272630) was that we have until July 14 to do all the final iterations. 

Regardless, I was planning to take a look at this today and will submit updates based on the review comments within a few days at the latest.

jpivarski commented:
My email said,

> If you agree, please reply to this email with your GitHub username by
Wednesday, June 7th. We ask that you complete an initial review no
later than June 23rd, and to work with the author through July 14th,
or until you feel the paper is ready for inclusion in the proceedings.
More detailed instructions can be found in the README
(https://github.com/scipy-conference/scipy_proceedings#instructions-for-reviewers).

and that README says,

> ### Reviewer Deadlines
> 
> - June 2: Reviewers Assigned
> - June 23: Initial Complete Review
> - July 21: Final Recommendation and Comprehensive Review Deadlines

which I interpret as "June 23 is when _reviewers_ need to provide a review to the _authors_." I only did it early because I had a vacation that covered that date. (I'm back now.)

So unless I'm misinterpreting it, you're not late.

cbcunc commented:
@scoobies mark pending comment

asmeurer committed:
Update the array API paper according to the review comments

asmeurer commented:
@cbcunc @jpivarski we've updated the paper according to your review comments.

jpivarski commented:
I re-read the paper and it looks good to me! Then I also scanned through the differences in (3cf0735aac274af11f7b752d8ce6449cd324f682; all paper-updates in one commit) and they address my concerns.

I think the paper is ready to go!

kgryte commented:
Thanks for your thorough review @jpivarski!

saulshanabrook commented on papers/aaron_meurer/paper.rst:
```diff
@@ -512,12 +514,16 @@ individual elements, "operators" to perform logical and arithmetic operations
 data-dependent and thus difficult to implement in array libraries relying on
 static analysis for graph-based optimization.
 
+.. [#] The array API standard includes support for in-place operations via `__setitem__()`; however, behavior is undefined if an in-place operation would affect arrays other than the target array (e.g., in array libraries supporting multiple "views" of the same underlying memory).
```
Thanks for this clarification! I was just wondering about this.

cbcunc commented:
@asmeurer, @jpivarski Sorry, I meant to say initial review deadline. I also somehow was looking at an old schedule. Thank you for the review and thank you for addressing the review comments. This is my favorite paper of this year's crop.

cbcunc commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-4af1f417-00a is OK
- 10.48550/arXiv.1209.5145 is OK
- 10.1137/141000671 is OK
- 10.48550/arXiv.1512.01274 is OK
- 10.1145/42288.42291 is OK
- 10.1145/77626.79170 is OK
- 10.1063/1.4822400 is OK
- 10.1109/MCSE.2007.51 is OK
- 10.1007/s00355-011-0603-9 is OK
- 10.1080/10361146.2014.900530 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.5334/jors.148 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1109/IEEESTD.2019.8766229 is OK
- 10.3233/978-1-61499-649-1-87 is OK
- 10.1145/355841.355847 is OK
- 10.21105/joss.01891 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.1002/0471725293 is OK
- 10.1109/MCSE.2011.36 is OK
- 10.1145/3386331 is OK
- 10.1109/MCSE.2007.58 is OK
- 10.1109/MCSE.2007.53 is OK
- 10.1109/MCSE.2010.119 is OK
- 10.1201/9781315373461-6 is OK
- 10.25080/Majora-7b98e3ed-013 is OK
- 10.25080/Majora-92bf1922-011 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.7717/peerj.453 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1109/TAU.1967.1161901 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.5555/3026877.3026899 is INVALID
- 10.1.1.112.9899 is INVALID
- 10.5555/3454287.3455008 is INVALID
- 10.5555/1953048.2078195 is INVALID
```

cbcunc commented:
@asmeurer would you please address the invalid DOIs above?

cbcunc commented:
@jpivarski does this paper meet your final review approval?

jpivarski commented:
> @jpivarski does this paper meet your final review approval?

Yes, I took another look at it, and it looks nice!

asmeurer commented:
> ```
> - 10.5555/3026877.3026899 is INVALID
> - 10.5555/3454287.3455008 is INVALID
> - 10.5555/1953048.2078195 is INVALID
> ```

I'm not sure what's up with DOIs, but these seem to come from ACM:

https://dl.acm.org/doi/10.5555/3026877.3026899
https://dl.acm.org/doi/10.5555/3454287.3455008
https://dl.acm.org/doi/10.5555/1953048.2078195

But they aren't resolvable by https://dx.doi.org/. I guess doi.org doesn't index ACM DOIs?

> ```
> - 10.1.1.112.9899 is INVALID
> ```

This one comes from the URL for the paper, https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.112.9899, but it doesn't even fit the format of a DOI correctly. I don't know if there is a real DOI for this paper.

cbcunc commented:
@asmeurer 

10.5555 is an indication of a fake DOI:

https://www.crossref.org/blog/doi-like-strings-and-fake-dois/

I think you are right about ACM. In that case, would you please simply remove the DOI metadata for the for the invalid DOI references and we'll just have to accept the references without DOI. I think that numarray paper is so old that I wouldn't expect it to have a DOI anyway.

Thank you for your attention and as soon as this is taken care of, the paper will be marked ready for publication.

asmeurer committed:
Remove invalid DOIs from the bibtex file

asmeurer commented:
I've removed those DOIs from the bibtex file.

cbcunc commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.25080/Majora-4af1f417-00a is OK
- 10.48550/arXiv.1209.5145 is OK
- 10.1137/141000671 is OK
- 10.48550/arXiv.1512.01274 is OK
- 10.1145/42288.42291 is OK
- 10.1145/77626.79170 is OK
- 10.1063/1.4822400 is OK
- 10.1109/MCSE.2007.51 is OK
- 10.1007/s00355-011-0603-9 is OK
- 10.1080/10361146.2014.900530 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.5334/jors.148 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1109/IEEESTD.2019.8766229 is OK
- 10.3233/978-1-61499-649-1-87 is OK
- 10.1145/355841.355847 is OK
- 10.21105/joss.01891 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.1002/0471725293 is OK
- 10.1109/MCSE.2011.36 is OK
- 10.1145/3386331 is OK
- 10.1109/MCSE.2007.58 is OK
- 10.1109/MCSE.2007.53 is OK
- 10.1109/MCSE.2010.119 is OK
- 10.1201/9781315373461-6 is OK
- 10.25080/Majora-7b98e3ed-013 is OK
- 10.25080/Majora-92bf1922-011 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.7717/peerj.453 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1109/TAU.1967.1161901 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

cbcunc commented:
@scoobies mark ready

deniederhut reviewed (approved):
Thanks @jpivarski for helping us get this paper ready for publication! And thanks @asmeurer for a truly phenomenal paper!

![source-191350742](https://github.com/scipy-conference/scipy_proceedings/assets/6868396/ff56266c-7cbd-4a40-b1ac-5f33ee1c1e6a)
